### ISSUE TITLE
[Bug] ImageClassificationInferencer and ImageClassificationInferencer.model produce different outputs

### ISSUE THREAD

**Issue by tjjack1226** [open-mmlab/mmpretrain#1684]
### Branch

main branch (mmpretrain version)

### Describe the bug

Using the mmpretrain.api.ImageClassificationInferencer is giving me outputs that are inconsistent with the underlying torch model.  I have a simple trained ResNet18, GAP and LinearCls head model with 3 classes and the default data_preprocessor (mean and std normalization), but when I run the following code, I do not get the same output predicted scores:

```
from mmpretrain.apis import ImageClassificationInferencer
import cv2
import torch

config_path = './path/to/config.py'
checkpoint_path = './path/to/checkpoint.pth'
inferencer = ImageClassificationInferencer(config_path, checkpoint_path, 'cpu')

img_name = './path/to/image.png'
bgr_img = cv2.imread(img_name, -1)
print(inferencer(bgr_img))
#example output: [{'pred_scores': array([2.2630255e-01, 7.7339667e-01, 3.0080788e-04], dtype=float32), 'pred_label': 1, 'pred_score': 0.7733966708183289, 'pred_class': <class 1>}]

rgb_img = cv2.cvtColor(bgr_img, cv2.COLOR_BGR2RGB)
#input is (B, C, H, W)
rgb_tensor = torch.from_numpy(rgb_img).permute(2,0,1).unsqueeze(0).float() 
#perform normalization
rgb_normalized = (rgb_tensor - inferencer.model.data_preprocessor.mean)/inferencer.model.data_preprocessor.std 
print(inferencer.model(rgb_normalized))
#example output: tensor([[ 1.2111,  2.4400, -5.4120]], grad_fn=<AddmmBackward0>)
```

I would think that the pred_scores array from the first print line should be identifcal to the output of the second print line, but they are very different.  

### Environment

 'sys.platform': 'win32',
 'Python': '3.9.16 (main, May 17 2023, 17:49:16) [MSC v.1916 64 bit (AMD64)]',
 'CUDA available': True,
 'numpy_random_seed': 2147483648,
 'GPU 0': 'NVIDIA GeForce RTX 3090 Ti',
 'CUDA_HOME': 'C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v11.7',
 'NVCC': 'Cuda compilation tools, release 11.7, V11.7.64',
 'MSVC': 'Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29924 for x64',
 'GCC': 'n/a',
 'PyTorch': '2.0.1+cu117',
 'TorchVision': '0.15.2+cu117',
 'OpenCV': '4.6.0',
 'MMEngine': '0.7.4',
 'MMCV': '2.0.0',
 'MMPreTrain': '1.0.0rc8+1c7a700'

### Other information

I've not made any modifications to the code.  I'm using my own config file using the default ResNet18 one as a base.  The only ops in my test_pipeline are LoadImageFromFile and PackInputs.  I've also attempted to use mmdeploy.apis.build_task_processor to build_pytorch_model, but get the same inconsistencies.

**Comment by Ezra-Yu** [comment]
I don't think you've fully aligned the data pre-processing. there should be a pipeline of `Resize` in the pre-process.

**Comment by tjjack1226** [comment]
Thank you for your response.  My first guess was also that there was a size mismatch, but I'm pretty sure that's not the case.  I do have a `dict(type='Resize', size=(360,640), backend='pillow')` in the config, but the input image in the above example is already that size so not necessary.  Adding a call to `PIL.Image.resize` does not change the output.  There is also no padding occuring in my preprocessing, as the `ClsDataPreprocessor` padding parameters are kept at their default values.  Unless there's also some resizing inside of `ClsDataPreprocessor` that I've not accounted for?

**Comment by tjjack1226** [comment]
For what it's worth, I've also tried to use mmdeploy's `build_task_processor` and `task_processor.create_input` to generate preprocessed inputs using my exact config file to then pass into `ImageClassificationInferencer.model`.  When I do that, I get results that are identical to my `inferencer.model(rgb_normalized)` code in the original post, which is still different from calling `ImageClassificationInferencer` directly.

**Comment by tjjack1226** [comment]
Turned out the problem seemed to be the default argument `mode` of `inferencer.model.forward`.  Looks like the default `mode` is `"tensor"` but in order to recreate the results, I need it set to `"predict"`.  Any idea why `"tensor"` and `"predict"` are generating different results?  

In any case, I was able to use `functools.partial` to change the mode and make the results identical.  Posting in case this helps anyone else, as it was critical for me to solve in order to be able to export my model to onnx:

```
from mmpretrain.apis import ImageClassificationInferencer
import cv2
import torch
from functools import partial

config_path = './path/to/config.py'
checkpoint_path = './path/to/checkpoint.pth'
img_name = './path/to/image.png'

inferencer = ImageClassificationInferencer(config_path, checkpoint_path, 'cpu')
bgr_img = cv2.imread(img_name, -1)
print(inferencer(bgr_img))
#example output: [{'pred_scores': array([2.2630255e-01, 7.7339667e-01, 3.0080788e-04], dtype=float32), 'pred_label': 1, 'pred_score': 0.7733966708183289, 'pred_class': <class 1>}]

rgb_img = cv2.cvtColor(bgr_img, cv2.COLOR_BGR2RGB)
rgb_tensor = torch.from_numpy(rgb_img).permute(2,0,1).unsqueeze(0).float()
rgb_normalized = (rgb_tensor - inferencer.model.data_preprocessor.mean)/inferencer.model.data_preprocessor.std

#change mode argument with partial
inferencer.model.forward = partial(inferencer.model.forward, mode='predict')
print(inferencer.model(rgb_normalized)[0].pred_score)`
#example output: tensor([2.2630e-01, 7.7340e-01, 3.0081e-04], grad_fn=<UnbindBackward0>)
```
